rust/osm2lanes/src/tags.rs

```diff
@@ -0,0 +1,141 @@
+use serde::Deserialize;
+
+use std::collections::BTreeMap;
+
+/// A representation for a OSM tags key
+#[derive(Clone)]
+pub enum TagKey {
+    Static(&'static str),
+    String(String),
+}
+
+impl TagKey {
+    pub const fn from(string: &'static str) -> Self {
+        TagKey::Static(string)
+    }
+    pub fn as_str(&self) -> &str {
+        match self {
+            Self::Static(v) => v,
+            Self::String(v) => v.as_str(),
+        }
+    }
+}
+
+impl From<&'static str> for TagKey {
+    fn from(string: &'static str) -> Self {
+        TagKey::from(string)
+    }
+}
+
+impl std::ops::Add for TagKey {
+    type Output = Self;
+    fn add(self, other: Self) -> Self {
+        let val = format!("{}:{}", self.as_str(), other.as_str());
+        TagKey::String(val)
+    }
+}
+
+impl std::ops::Add<&'static str> for TagKey {
+    type Output = Self;
+    fn add(self, other: &'static str) -> Self {
+        self.add(TagKey::from(other))
+    }
+}
+
+/// A map from string keys to string values. This makes copies of strings for
+/// convenience; don't use in performance sensitive contexts.
+// BTreeMap chosen for deterministic serialization.
+// We often need to compare output directly, so cannot tolerate reordering
+// TODO: fix this in the serialization by having the keys sorted.
+#[derive(Clone, Deserialize, Default)]
+pub struct Tags(BTreeMap<String, String>);
+
+impl Tags {
+    pub fn new(map: BTreeMap<String, String>) -> Tags {
+        Tags(map)
+    }
+
+    /// Expose inner map
+    pub fn map(&self) -> &BTreeMap<String, String> {
+        &self.0
+    }
+}
+
+impl std::str::FromStr for Tags {
+    type Err = String;
+
+    /// Parse tags from an '=' separated list
+    ///
+    /// ```
+    /// use std::str::FromStr;
+    /// use osm2lanes::Tags;
+    /// use osm2lanes::TagsRead;
+    /// let tags = Tags::from_str("foo=bar\nabra=cadabra").unwrap();
+    /// assert_eq!(tags.get("foo"), Some("bar"));
+    /// ```
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut map = BTreeMap::new();
+        for line in s.lines() {
+            let (key, val) = line.split_once("=").ok_or("tag must be = separated")?;
+            map.insert(key.to_owned(), val.to_owned());
+        }
+        Ok(Self(map))
+    }
+}
+
+// TODO, shouldn't TagKey be passed by reference?
```

Comment:
Yeah, that's what I'd expect. If you just switch from `Into` to [AsRef](https://doc.rust-lang.org/std/convert/trait.AsRef.html), that may work?

Reply:
Hmm, not straightforward, not sure and have to get back to other work..